src/Microsoft.Azure.SignalR.Common/Utilities/BackOffPolicy.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.SignalR.Common
+{
+    internal class BackOffPolicy
+    {
+        private TaskCompletionSource<bool> _currentProbeTcs = null;
+        private int _currentRetryCount = 0;
+
+        /// <summary>
+        /// Provides a synchronized mechanism of calling probing funcs by multiple concurrent callers.
+        /// Each caller's probe func will be invoked exactly one time.
+        /// The probe call may get delayed depending on the result of previous probe calls.
+        /// The delay is controlled by getRetryDelay func and the number of consecutive failed probe calls.
+        /// </summary>
+        /// <param name="probe"> this func returns a task with boolean result indicating if the probe was successful</param>
+        /// <param name="getRetryDelay"> this func returns a TimeSpan delay for a given iteration number</param>
+        /// <returns> 
+        /// A task with its final state and result matching the state and result of the task returned by the probe func.
+        /// This task will completes after either of the following happens:
+        /// - probe's task changes state to task.IsCompletedSuccessfully == true and task.Result == true
+        /// - probe's task.Result == false / faulted / cancelled and the delay defined by getRetryDelay has passed
+        /// </returns>
+        public async Task<bool> CallProbeWithBackOffAsync(Func<Task<bool>> probe, Func<int, TimeSpan> getRetryDelay)
+        {
+            bool calledProbeOnce = false;
+            bool probeSuccess = false;
+            do
+            {
+                var myTcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+                // ensure only one caller will be selected to call probing func
+                var ongoingProbeTcs = Interlocked.CompareExchange(ref _currentProbeTcs, myTcs, null);
+                if (ongoingProbeTcs == null)
+                {
+                    // initiate the probe and indicate its result to others
+                    Task<bool> probeTask = null;
+                    bool awaitProbeTask = false;
+                    try
+                    {
+                        Debug.Assert(!calledProbeOnce);
+                        calledProbeOnce = true;
+                        probeTask = probe();
+
+                        using (CancellationTokenSource delayCts = new CancellationTokenSource())
+                        {
+                            var d = getRetryDelay(_currentRetryCount++);
+                            var delayTask = Task.Delay(d, delayCts.Token);
+                            await Task.WhenAny(delayTask, probeTask);
```

Comment:
Why isn't this looking at which task caused the yield?

Reply:
Even if it does we'd still need to handle 3 cases (success, timeout, failure) separately so this doesn't provide much benefit in this particular case.

Reply:
I dunno, looking at the code, it feels like it could be simplified

Reply:
I've seen both styles in .net libraries. The main goal was to keep the logic relatively easy to follow (successful connection, timeout, everything else).